## Homework I (Group of Five or less): Linear Regression

### 1. The tensile strength of a paper product is related to the amount of hardwood in the pulp. Ten samples are produced in the pilot plant, and the data obtained are shown in the following table.

In [3]:
import pandas as pd
tensile_dict = {'Strength' : [160, 171, 175, 182, 184, 181, 188, 193, 195, 200], 
                'Percent_Hardwood' : [10, 15, 15, 20, 20, 20, 25, 25, 28, 30] }
tensile_data = pd.DataFrame(data=tensile_dict)
tensile_data.head(10)

Strength  Percent_Hardwood
0       160                10
1       171                15
2       175                15
3       182                20
4       184                20
5       181                20
6       188                25
7       193                25
8       195                28
9       200                30

### (a) Fit a linear regression model relating strength topercent hardwood.

In [4]:
from statsmodels.formula.api import ols

formula = 'Strength ~ Percent_Hardwood'
#formula = 'Strength ~ Percent_Hardwood - 1'

tensile_lm = ols(formula, data=tensile_data).fit()
tensile_lm.params

Intercept           143.824385
Percent_Hardwood      1.878635
dtype: float64

### (b) Test the model in part (a) for significance of regression.

In [5]:
tensile_lm.summary2()

C:\Users\USER\AppData\Roaming\Python\Python38\site-packages\scipy\stats\stats.py:1541: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=10
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


<class 'statsmodels.iolib.summary2.Summary'>
"""
                  Results: Ordinary least squares
===================================================================
Model:                OLS              Adj. R-squared:     0.966   
Dependent Variable:   Strength         AIC:                45.9456 
Date:                 2021-09-11 15:49 BIC:                46.5507 
No. Observations:     10               Log-Likelihood:     -20.973 
Df Model:             1                F-statistic:        260.0   
Df Residuals:         8                Prob (F-statistic): 2.20e-07
R-squared:            0.970            Scale:              4.8541  
-------------------------------------------------------------------
                  Coef.   Std.Err.    t    P>|t|   [0.025   0.975] 
-------------------------------------------------------------------
Intercept        143.8244   2.5215 57.0386 0.0000 138.0097 149.6390
Percent_Hardwood   1.8786   0.1165 16.1245 0.0000   1.6100   2.1473
-------------------------------------------------------------------
Omnibus:                1.211        Durbin-Watson:           2.132
Prob(Omnibus):          0.546        Jarque-Bera (JB):        0.701
Skew:                   0.157        Prob(JB):                0.704
Kurtosis:               1.742        Condition No.:           78   
===================================================================

"""

### จะเห็นว่า Prob (F-statistic) มีค่าเท่ากับ 2.20e-07 ซึ่งมีค่าน้อยกว่า 0.05 จึงแสดงว่า Model นี้ significance

### (c) Find a 95 percent confidence interval on the parameter Beta1

### 95 percent confidence interval on the parameter Beta1 มีค่า 1.6100 ถึง	2.1473

### 2. A sample of 25 brands of cigarette has been tested for tar and nicotine contents, carbonmonoxide (mgs) emitted by cigarette smoke, and cigarette weight (g). Past studies have shown that the amount CO emitted increases with the tar and nicotine content.

In [6]:
co_data = pd.read_excel('data/lect02-lin-reg.xlsx', sheet_name='CO-smoke', header=0)
co_data.head()

TAR  NICOTINE  WEIGHT    CO
0  14.1      0.86  0.9853  13.6
1  16.0      1.06  1.0938  16.6
2  29.8      2.03  1.1650  23.5
3   8.0      0.67  0.9280  10.2
4   4.1      0.40  0.9462   5.4

### a) Model the CO emitted by cigratte as a linear function of the other three quantities. Determine SSR, SSE, R?, adjusted R2, and show the fitted model coefficients.

In [9]:
from statsmodels.formula.api import ols

#formula = 'CO ~ TAR + NICOTINE + WEIGHT'
formula = 'CO ~ TAR + NICOTINE + WEIGHT - 1'

co_lm = ols(formula, data=co_data).fit()
co_lm.params

TAR         0.936701
NICOTINE   -2.495939
WEIGHT      3.349901
dtype: float64

In [22]:
print("Sum of Squared Regression (SSR): {:.2f}".format(co_lm.ess)) # Explained sum of squares
print("Sum of Squared Errors (SSE): {:.2f}".format(co_lm.ssr)) # Sum of Squares Residuals
print("R-squared: {:.3f}".format(co_lm.rsquared))
print("Adjusted R-squared: {:.3f}".format(co_lm.rsquared_adj))

Sum of Squared Regression (SSR): 4417.24
Sum of Squared Errors (SSE): 45.68
R-squared: 0.990
Adjusted R-squared: 0.988


### b) Which variables are significant and not significant in the model ?

In [10]:
co_lm.summary2()

<class 'statsmodels.iolib.summary2.Summary'>
"""
                       Results: Ordinary least squares
==============================================================================
Model:                  OLS              Adj. R-squared (uncentered): 0.988   
Dependent Variable:     CO               AIC:                         92.0171 
Date:                   2021-09-11 15:50 BIC:                         95.6737 
No. Observations:       25               Log-Likelihood:              -43.009 
Df Model:               3                F-statistic:                 709.1   
Df Residuals:           22               Prob (F-statistic):          4.97e-22
R-squared (uncentered): 0.990            Scale:                       2.0764  
----------------------------------------------------------------------------------
                Coef.      Std.Err.        t        P>|t|       [0.025      0.975]
----------------------------------------------------------------------------------
TAR             0.9367       0.2398      3.9056     0.0008       0.4393     1.4341
NICOTINE       -2.4959       3.8850     -0.6425     0.5272     -10.5529     5.5610
WEIGHT          3.3499       0.9661      3.4675     0.0022       1.3464     5.3534
------------------------------------------------------------------------------
Omnibus:                  0.689             Durbin-Watson:               2.938
Prob(Omnibus):            0.708             Jarque-Bera (JB):            0.380
Skew:                     -0.298            Prob(JB):                    0.827
Kurtosis:                 2.896             Condition No.:               184  
==============================================================================

"""

#### significant variables : Tar, Weight Not significant variables : Nicotine

### c)  Remove the most significant variable in a) and refit the model. How do the model coefficients and your conclusion about the significance of variables change compared to a)?

In [24]:
from statsmodels.formula.api import ols

#formula = 'CO ~ TAR + NICOTINE + WEIGHT'
formula = 'CO ~ NICOTINE + WEIGHT - 1'

co_lm = ols(formula, data=co_data).fit()
co_lm.params

NICOTINE    12.251221
WEIGHT       1.834121
dtype: float64

In [25]:
co_lm.summary2()

<class 'statsmodels.iolib.summary2.Summary'>
"""
                       Results: Ordinary least squares
==============================================================================
Model:                  OLS              Adj. R-squared (uncentered): 0.981   
Dependent Variable:     CO               AIC:                         103.1850
Date:                   2021-08-29 13:23 BIC:                         105.6227
No. Observations:       25               Log-Likelihood:              -49.592 
Df Model:               2                F-statistic:                 652.0   
Df Residuals:           23               Prob (F-statistic):          5.58e-21
R-squared (uncentered): 0.983            Scale:                       3.3632  
----------------------------------------------------------------------------------
                Coef.      Std.Err.        t        P>|t|       [0.025      0.975]
----------------------------------------------------------------------------------
NICOTINE       12.2512       1.1636     10.5286     0.0000      9.8441     14.6583
WEIGHT          1.8341       1.1260      1.6289     0.1169     -0.4951      4.1633
------------------------------------------------------------------------------
Omnibus:                   0.142            Durbin-Watson:               2.747
Prob(Omnibus):             0.932            Jarque-Bera (JB):            0.340
Skew:                      0.115            Prob(JB):                    0.844
Kurtosis:                  2.477            Condition No.:               6    
==============================================================================

"""

#### จากโมเดล A มี variable 2 ตัวที่ Significant คือ Tar กับ Weight และตัวที่ Not significant คือ Nicotine ซึ่งมีค่า coefficients เท่ากับ -2.4959 แต่เมื่อเปลี่ยนโมเดลใหม่โดยลบ most significant variable ซึ่งคือ Tar ออกไปทำให้ Nicotine กลายเป็น most significant variable ซึ่งมีค่า coefficients เพิ่มขึ้นจากโมเดล A